src/bsp/platform/esp32/xi_bsp_io_fs_esp32.c

```diff
@@ -0,0 +1,307 @@
+
+    if ( NULL == resource_name || NULL == resource_handle_out )
+    {
+        return XI_BSP_IO_FS_INVALID_PARAMETER;
+    }
+
+    if ( xi_bsp_fwu_is_this_firmware( resource_name ) )
+    {
+        const esp_partition_t* next_partition = esp_ota_get_next_update_partition( NULL );
+
+        if ( NULL == next_partition )
+        {
+            xi_bsp_debug_logger( "Failed to get next OTA update partition" );
+            return XI_BSP_IO_FS_OPEN_ERROR;
+        }
+
+        esp_retv = esp_ota_begin( next_partition, size,
+                                  ( esp_ota_handle_t* )resource_handle_out );
+
+        if ( ESP_OK != esp_retv )
+        {
+            xi_bsp_debug_format( "esp_ota_begin() failed with error %d", esp_retv );
+            return XI_BSP_IO_FS_OPEN_ERROR;
+        }
+
+        esp32_xibsp_notify_update_started( resource_name, size );
+        open_firmware_bin_handle = *resource_handle_out;
+    }
+    else
+    {
+        snprintf( filepath, XI_ESP32_MAX_FILENAME_LEN, "%s/%s", ESP32_FS_BASE_PATH,
+                  resource_name );
+        fp = fopen( ( const char* )filepath,
+                    ( open_flags & XI_BSP_IO_FS_OPEN_READ ) ? "rb" : "wb" );
+
+        /* if error on fopen check the errno value */
+        XI_BSP_IO_FS_CHECK_CND(
+            NULL == fp, xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( errno ), ret );
+
+        *resource_handle_out = ( xi_bsp_io_fs_resource_handle_t )fp;
+
+        return XI_BSP_IO_FS_STATE_OK;
+    }
+
+    return XI_BSP_IO_FS_STATE_OK;
+
+err_handling:
+    return ret;
+}
+
+xi_bsp_io_fs_state_t
+xi_bsp_io_fs_write( const xi_bsp_io_fs_resource_handle_t resource_handle,
+                    const uint8_t* const buffer,
+                    const size_t buffer_size,
+                    const size_t offset,
+                    size_t* const bytes_written )
+{
+    ( void )offset;
+    xi_bsp_io_fs_state_t ret = XI_BSP_IO_FS_STATE_OK;
+    esp_err_t retv           = ESP_OK;
+
+    if ( ( esp_ota_handle_t )resource_handle == open_firmware_bin_handle )
+    {
+        retv = esp_ota_write( resource_handle, buffer, buffer_size );
+        if ( ESP_OK != retv )
+        {
+            xi_bsp_debug_format( "esp_ota_write() failed with error %d", retv );
+            return XI_BSP_IO_FS_WRITE_ERROR;
+        }
+
+        *bytes_written = buffer_size;
+        esp32_xibsp_notify_chunk_written( buffer_size, offset );
+    }
+    else
+    {
+        if ( NULL == buffer || 0 == buffer_size ||
+             XI_BSP_IO_FS_INVALID_RESOURCE_HANDLE == resource_handle )
+        {
+            return XI_BSP_IO_FS_INVALID_PARAMETER;
+        }
+
+        FILE* fp = ( FILE* )resource_handle;
+
+        /* let's set the offset */
+        const int fop_ret = fseek( fp, offset, SEEK_SET );
+
+        /* if error on fseek check errno */
+        XI_BSP_IO_FS_CHECK_CND(
+            fop_ret != 0, xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( errno ), ret );
```

Comment:
have you tested the non firmware file update?

Reply:
Good point... I've only tested it with `.xirev` files, not regular updates. I'm gonna test it right now....